### PR TITLE
Expose a connection limit option

### PR DESCRIPTION
**Pull request by Tratcher**
#519

### CODE REVIEW

File: test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/ServerTests.cs [hunk 1]
```diff
@@ -300,6 +300,63 @@ public void Server_MultipleOutstandingAsyncRequests_Success()
...
+        [ConditionalFact]
+        public async Task Server_SetConnectionLimitInfinite_Success()
+        {
+            // This is just to get a dynamic port
+            string address;
+            using (Utilities.CreateHttpServer(out address, httpContext => Task.FromResult(0))) { }
+
+            var server = Utilities.CreatePump();
+            server.Listener.Options.UrlPrefixes.Add(UrlPrefix.Create(address));
+            server.Listener.Options.MaxConnections = Timeout.Infinite; // -1
```
Comment (davidfowl) [inline test/Microsoft.AspNetCore.Server.HttpSys.FunctionalTests/ServerTests.cs]: Why not just use -1? Using Timeout.Infitinite is obscure.

File: src/Microsoft.AspNetCore.Server.HttpSys/NativeInterop/UrlGroup.cs [hunk 2]
```diff
@@ -33,6 +37,19 @@ internal unsafe UrlGroup(ServerSession serverSession, ILogger logger)
 
         internal ulong Id { get; private set; }
 
+        internal unsafe void SetMaxConnections(long maxConnections)
+        {
+            var connectionLimit = new HttpApi.HTTP_CONNECTION_LIMIT_INFO();
+            connectionLimit.Flags = HttpApi.HTTP_FLAGS.HTTP_PROPERTY_FLAG_PRESENT;
+            connectionLimit.MaxConnections = (uint)maxConnections;
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.HttpSys/NativeInterop/UrlGroup.cs]: How convenient.

Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.HttpSys/NativeInterop/UrlGroup.cs]: Yeah, this is what IIS uses too. It wouldn't be possible otherwise.

File: src/Microsoft.AspNetCore.Server.HttpSys/HttpSysOptions.cs [hunk 3]
```diff
@@ -55,6 +57,28 @@ public HttpSysOptions()
         public bool ThrowWriteExceptions { get; set; }
 
         /// <summary>
+        /// Gets or Sets the maximum number of concurrent connections to accept, -1 for infinate, or null for unspecified.
```
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Server.HttpSys/HttpSysOptions.cs]: What is the effective difference between unspecified and infinite? In Kestrel, we use null to mean infinite.

typos: "Gets or sets", "infinite".

Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.HttpSys/HttpSysOptions.cs]: unspecified means to use the OS default. many http.sys settings can be configured system wide in the registry.
https://support.microsoft.com/en-us/help/820129/http.sys-registry-settings-for-windows

File: src/Microsoft.AspNetCore.Server.HttpSys/HttpSysOptions.cs [hunk 4]
```diff
@@ -55,6 +57,29 @@ public HttpSysOptions()
         public bool ThrowWriteExceptions { get; set; }
 
         /// <summary>
+        /// Gets or sets the maximum number of concurrent connections to accept, -1 for infinite, or null to
+        /// use the machine wide setting from the registry. The default value is null.
+        /// </summary>
+        public long? MaxConnections
```
Comment (natemcmaster) [inline src/Microsoft.AspNetCore.Server.HttpSys/HttpSysOptions.cs]: Would be good to add tests asserting the default value = null, that users can set valid values, and the out of range exception.

Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.HttpSys/HttpSysOptions.cs]: Added
